chore: detect os/arch and warn if ARM arch

huanjani opened the pull request:
This change logs a warning and suggests a workaround for ARM machines. 

Tested that the warning was logged on ARM arch, and also that `svc deploy` worked with the workaround.
Next step: target platform amd64/x86 for ARM builds.

(This includes an unrelated addition to the Makefile.)

Related: #1949, #2369.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

huanjani committed:
build: detect os/arch of docker engine; error out if ARM

huanjani committed:
Merge remote-tracking branch 'archer/mainline' into detect

huanjani committed:
chore: add unit tests, minor refactoring, add Makefile mocks file

huanjani committed:
chore: change error to log statement

huanjani committed:
chore: change from infof to warningln, needed newline

huanjani committed:
chore: change to warning f, update comment

huanjani committed:
chore: tweak formatting to make command more readable

huanjani commented:
manual testing output:
```
$ copilot svc init
Note: It's best to run this command in the root of your workspace.
Service type: Load Balanced Web Service
Service name: arm-ed-service
Dockerfile: ./Dockerfile
no EXPOSE statements in Dockerfile ./Dockerfile
Port: 80
Note: Architecture type arm64 is currently unsupported. To deploy, run `DOCKER_DEFAULT_PLATFORM=linux/amd64 copilot svc deploy`
✔ Wrote the manifest for service arm-ed-service at copilot/arm-ed-service/manifest.yml
Your manifest contains configurations like your container size and port (:80).

⠇ Creating ECR repositories for service arm-ed-service. 
```

huanjani committed:
fix: add quotes to yaml

iamhopaul123 commented on internal/pkg/exec/docker.go:
```diff
@@ -166,6 +166,36 @@ func (c DockerCommand) CheckDockerEngineRunning() error {
+		return "", "", ErrDockerCommandNotFound
+	}
+	osBuf := &bytes.Buffer{}
+	err = c.runner.Run("docker", []string{"version", "-f", "'{{.Server.Os}}'"}, Stdout(osBuf))
```
Stupid question: why do we use `Server` instead of `Client`?

huanjani replied:
Good question! Because the user could be on a Windows machine running WSL, for example, in which case the host os/arch would be different than the engine os/arch, and it's the engine one we're interested in because that will match the container platform.

iamhopaul123 commented on internal/pkg/cli/svc_init.go:
```diff
@@ -220,13 +222,22 @@ func (o *initSvcOpts) Execute() error {
 		}
 	}
 
+	o.os, o.arch, err = o.getOSArch()
```
hmm what if they don't have a docker daemon running? Will it break since I guess they can use Copilot without docker running with existing image?

huanjani replied:
Yes. 
It errors out with : 
```
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
✘ get os/arch from docker: get docker os: exit status 1
```
But you're right-- they should be allowed to do this. Thanks for catching this. Will fix!

huanjani committed:
chore: add platform functionality to job init

huanjani committed:
chore: use GOOS and GOARCH if no daemon running

efekarakus commented on internal/pkg/cli/job_init.go:
```diff
@@ -323,6 +335,20 @@ func jobTypePromptOpts() []prompt.Option {
 	return options
 }
 
+func (o initJobOpts) getOSArch() (os, arch string, err error) {
```
```suggestion
func (o *initJobOpts) dockerPlatform() (os, arch string, err error) {
```

efekarakus commented on internal/pkg/cli/job_init.go:
```diff
@@ -323,6 +335,20 @@ func jobTypePromptOpts() []prompt.Option {
+		return "", "", fmt.Errorf("get os/arch from docker: %w", err)
+	}
+	// Until we target X86_64 for ARM architectures, log a warning.
+	if arch == "arm" || arch == "arm64" {
```
Should we store these in constants somewhere?

I think it would make sense for us to refactor in a separate PR `exec` and create two separate packages: `docker/dockerengine` and `docker/dockerfile` and shove these constants to `dockerengine`

efekarakus commented on internal/pkg/cli/job_init.go:
```diff
@@ -323,6 +335,20 @@ func jobTypePromptOpts() []prompt.Option {
+	}
+	// Until we target X86_64 for ARM architectures, log a warning.
+	if arch == "arm" || arch == "arm64" {
+		log.Warningf("Architecture type %s is currently unsupported. To deploy, run %s\n", arch, "`DOCKER_DEFAULT_PLATFORM=linux/amd64 copilot svc deploy`")
```
nit:
```suggestion
		log.Warningf("Architecture type %s is currently unsupported.\nTo deploy, run %s\n", arch, "`DOCKER_DEFAULT_PLATFORM=linux/amd64 copilot deploy`")
```

efekarakus commented on internal/pkg/cli/svc_init.go:
```diff
@@ -420,6 +432,26 @@ func parseHealthCheck(df dockerfileParser) (*manifest.ContainerHealthCheck, erro
 }
 
+func (o initSvcOpts) getOSArch() (operatingSystem, architecture string, err error) {
+	var os string
+	var arch string
+	if o.image != "" {
+		os = runtime.GOOS
+		arch = runtime.GOARCH
```
Can we do something like this instead? 
```go
os, arch := "linux", "amd64"
if o.image == "" {
   os, arch, err = o.dockerEngineValidator.Platform()
}
```

huanjani replied:
If someone on an ARM machine supplies an existing Docker image that is for ARM architecture, they won't know why it is failing, though, right?

efekarakus replied:
oh that makes sense!
then let's do:
```go
os, arch := runtime.GOOS, runtime.GOARCH
if o.image == "" {
   ...
}
```

efekarakus commented on internal/pkg/cli/svc_init.go:
```diff
@@ -420,6 +432,26 @@ func parseHealthCheck(df dockerfileParser) (*manifest.ContainerHealthCheck, erro
+		os = runtime.GOOS
+		arch = runtime.GOARCH
+	} else {
+		os, arch, err = o.dockerEngineValidator.GetPlatform()
```
What do you think of renaming from`dockerEngineValidator` to just `dockerEngine`?

huanjani committed:
chore: use GOOS and GOARCH if docker daemon off and using image flag

efekarakus commented on internal/pkg/exec/docker.go:
```diff
@@ -166,6 +166,36 @@ func (c DockerCommand) CheckDockerEngineRunning() error {
 }
 
+// GetPlatform will run the `docker version` command to get the OS/Arch.
+func (c DockerCommand) GetPlatform() (operatingSystem, architecture string, err error) {
```
nit:
```suggestion
func (c DockerCommand) Platform() (os, arch string, err error) {
```

efekarakus commented on internal/pkg/exec/docker.go:
```diff
@@ -166,6 +166,36 @@ func (c DockerCommand) CheckDockerEngineRunning() error {
+	os = strings.TrimSuffix(strings.TrimPrefix(strings.TrimSpace(osBuf.String()), "'"), "'")
+
+	archBuf := &bytes.Buffer{}
+	err = c.runner.Run("docker", []string{"version", "-f", "'{{.Server.Arch}}'"}, Stdout(archBuf))
```
An alternative to issuing two commands:
```go
// Use the "json" function
buf := new(bytes.Buffer)
c.runner.Run("docker", []string{"version", "-f", "'{{json .Server}}'"}, Stdout(buf))

// unmarshal into data
daemon := struct {
  OS string `json:"Os"`
  Arch string `json:"Arch"`
}{}
json.Unmarshal(buf.Bytes(), &daemon)
```

huanjani replied:
that's what I had originally...will switch it back!

huanjani committed:
fix: switch darwin back to linux for check (darwin locally)

huanjani committed:
chore: tighten up naming

huanjani committed:
chore: combine Client:
 Cloud integration: 1.0.14
 Version:           20.10.6
 API version:       1.41
 Go version:        go1.16.3
 Git commit:        370c289
 Built:             Fri Apr  9 22:46:57 2021
 OS/Arch:           darwin/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.6
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       8728dd2
  Built:            Fri Apr  9 22:44:56 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.4
  GitCommit:        05f951a3781f4f2c1911b05e61c160e9c30eaa8e
 runc:
  Version:          1.0.0-rc93
  GitCommit:        12644e614e25b05da6fd08a38ffa0cfe1903fdec
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0 calls, replace if-else

huanjani committed:
fix: remove struct bc no longer need to check for non-empty response

huanjani commented:
**AMD64**
docker running (with or without image):
[no change]
docker not running, with image:
[no change]
docker not running, without image:
```
% copilot svc init                                                                               
Note: It's best to run this command in the root of your workspace.
Service type: Load Balanced Web Service
Service name: svc2
Docker daemon is not responsive; Copilot won't build from a Dockerfile.
Image: 
Port: 80
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
✘ get os/arch from docker: run docker version: exit status 1
```
**ARM**
docker running (with or without image):
```
$ copilot svc init
Note: It's best to run this command in the root of your workspace.
Service type: Load Balanced Web Service
Service name: svc2
Dockerfile: ./Dockerfile
no EXPOSE statements in Dockerfile ./Dockerfile
Port: 80
Note: Architecture type arm64 is currently unsupported.
To deploy, run `DOCKER_DEFAULT_PLATFORM=linux/amd64 copilot svc deploy`
✔ Manifest file for service svc2 already exists at copilot/svc2/manifest.yml, skipping writing it.
Your manifest contains configurations like your container size and port (:80).

⠸ Creating ECR repositories for service svc2. ✔ Created ECR repositories for service svc2.

Recommended follow-up actions:
- Update your manifest copilot/svc2/manifest.yml to change the defaults.
- Run `copilot svc deploy --name svc2 --env test` to deploy your service to a test environment.
```
docker not running, with image:
```
$ copilot svc init -i 123456789018.dkr.ecr.us-west-2.amazonaw
s.com/arm-based-repo/arm-ed-service
Note: It's best to run this command in the root of your workspace.
Service type: Load Balanced Web Service
Service name: svc2
Port: 80
Note: Architecture type arm64 is currently unsupported.
To deploy, run `DOCKER_DEFAULT_PLATFORM=linux/amd64 copilot svc deploy`
✔ Manifest file for service svc2 already exists at copilot/svc2/manifest.yml, skipping writing it.
Your manifest contains configurations like your container size and port (:80).

⠸ Creating ECR repositories for service svc2. ✔ Created ECR repositories for service svc2.

Recommended follow-up actions:
- Update your manifest copilot/svc2/manifest.yml to change the defaults.
- Run `copilot svc deploy --name svc2 --env test` to deploy your service to a test environment.
```
docker not running, without image:
```
$ copilot svc init
Note: It's best to run this command in the root of your workspace.
Service type: Load Balanced Web Service
Service name: svc2
Docker daemon is not responsive; Copilot won't build from a Dockerfile.
Image: 
Port: 80
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
✘ get os/arch from docker: run docker version: exit status 1
```

huanjani committed:
chore: make tests pass for whatever runtime platform

efekarakus commented on internal/pkg/cli/job_init.go:
```diff
@@ -323,6 +336,21 @@ func jobTypePromptOpts() []prompt.Option {
+	}
+	// Until we target X86_64 for ARM architectures, log a warning.
+	if arch == exec.ArmArch || arch == exec.Arm64Arch {
+		log.Warningf("Architecture type %s is currently unsupported.\nTo deploy, run %s\n", arch, "`DOCKER_DEFAULT_PLATFORM=linux/amd64 copilot svc deploy`")
```
nit: so that the command also works for jobs.
```suggestion
		log.Warningf("Architecture type %s is currently unsupported.\nTo deploy, run %s\n", arch, "`DOCKER_DEFAULT_PLATFORM=linux/amd64 copilot deploy`")
```

huanjani replied:
Removed "svc" from this warning in `svc_init` also.

efekarakus commented on internal/pkg/cli/svc_init.go:
```diff
@@ -420,6 +432,21 @@ func parseHealthCheck(df dockerfileParser) (*manifest.ContainerHealthCheck, erro
 	}, nil
 }
 
+func (o *initSvcOpts) dockerPlatform() (os, arch string, err error) {
```
This method is duplicated right? What do you think of refactoring it into a function:

```go
func dockerPlatform(engine dockerEngine) (os, arch string, err error)
```

efekarakus commented on internal/pkg/exec/docker.go:
```diff
@@ -48,8 +48,10 @@ type dockerConfig struct {
-const(
+const (
 	credStoreECRLogin = "ecr-login" // set on `credStore` attribute in docker configuration file
+	ArmArch           = "arm"
```
can we create a separate group and drop a comment here:
```go
// Architecture options given by Docker.
const (
  ArchARM = "arm"
  ArchARM64 = "arm64"
)
```

efekarakus commented on internal/pkg/exec/docker.go:
```diff
@@ -166,6 +168,33 @@ func (c DockerCommand) CheckDockerEngineRunning() error {
+	if err != nil {
+		return "", "", fmt.Errorf("run docker version: %w", err)
+	}
+	if c.buf != nil {
+		buf = c.buf
+	}
```
I don't think we need this for testing, we should be able to mock `runner.Run` and check if the output `os, arch` is as expected

efekarakus reviewed (approved):
Looks good! Minor refactor requests, feel free to remove the label once addressed :shipit:

huanjani committed:
Update internal/pkg/cli/job_init.go

Co-authored-by: Efe Karakus <efekarakus@gmail.com>

huanjani committed:
chore: turn method into function

huanjani committed:
fix: move Equal in unit test

huanjani committed:
refactor: make unit tests consistent with rest of package

iamhopaul123 reviewed (approved):
:shipit:

mergify[bot] committed:
Merge branch 'mainline' into detect
